python/dolfinx/fem/forms.py

```diff
@@ -333,3 +360,140 @@ def unique_spaces(V):
             return list(unique_spaces(V.transpose()))
 
     raise RuntimeError("Unsupported array of forms")
+
+
+@dataclass
+class CompiledForm:
+    """
+    Compiled UFL form without associated DOLFINx data
+
+    """
+
+    ufl_form: ufl.Form  # The original ufl form
+    ufcx_form: typing.Any  # The compiled form
+    module: typing.Any  #  The module
+    code: str  # The source code
+    dtype: npt.DTypeLike  # data type of coefficients and constants
```

Comment:
Is this needed? Surely the form itself already has a `dtype` and the constants and coefficients need to match that.

Reply:
You do not have the coefficients at compilation time (with dtype). You only have the abstract ufl objects.

Reply:
OK, but the `ufcx_form` is compiled for a particular `dtype`. Maybe we do need to store it here, but I think it should just be called the "dtype of the form". After all, you might not have any constants or coefficients.